docs/architecture/adr-022-system-merge.md

```diff
@@ -0,0 +1,63 @@
+
+- A bridge node was introduced which had to be run side by side with the consensus node, relaying data from the consensus node to the nodes in the DA network (i.e. using LibP2P). This meant:
+  - Extra complexity for node operators
+  - Duplicate storage of squares
+  - Redundant computation of the extended data square from the block
+  - Extra round trip in retrieving the data and pushing it (latency costs).
+- A secondary implementation of a storage engine which largely overlaps with the KV based storage engine in the consensus node.
+- A secondary implementation for exchanging and verifying data: `go-header`
+- Two binaries that have different usage patterns: commands, config files, flags etc. This makes it more confusing for users.
+
+Having duplication requires both a broader knowledge base and more maintenance load for engineers. There is a larger surface area for bugs and gains from optimizations are not captured as efficiently as they could.
+
+Outside of the aforementioned tech debt, there is untapped potential in taking the benefits of LibP2P and porting it across to the Tendermint consensus algorithm. LibP2P has a more sophisticated and efficient gossiping protocols than Tendermint's. It also has a more robust peer discovery and peer management system.
+
+## Decision
+
+The decision is to converge to a LibP2P based network and to shape out reusable components that can be used by all node types: consensus, data and light. This is to be gradually done across four phases, each phase likely requiring it's own in-depth ADR:
+
+### Phase 1: Removal of the bridge node
+
+The first step involves porting `Init` and `Start` methods from `celestia-app` and allowing consensus nodes to be initialized and run through the `celestia` binary. This will allow for a arguemnt to be passed in `Start` that enables the consensus nodes to directly push the extended data square (EDS) that is finalized in `FinalizeBlock` to the LibP2P based networks. Thus consensus nodes will temporarily be a hybrid of both networking systems and replace the need for a separate bridge node. Having a single binary will greatly simplify both the release process and end-to-end testing.
+
+### Phase 2: Migration of storage, syncing and fraud
+
+The next step migrates first the storage engine and then the syncing and fraud protocols (`go-header` and `go-fraud`) to `celestia-app`. This replaces the existing `BlockStore` and blocksync and evidence reactors. Streamlining the storage will require careful migration but will eliminate the duplication of storage and unify many of the data exchange protocols and the public APIs. Here, the consensus node can become more integrated with the `nodebuilder` package as it begins to be split up into a series of modules. This phase may also try to align configuration files and the general file directory.
```

Comment:
From what we discussed, we wanted the syncing and storage will be backed by Node's API. Something like Put/GetEDS, which Consensus uses to write and read edses, respectively. Similarly, we would provide getters and putters for Headers over API. I don't yet see how consensus would work with go-header directly. Also the API to trim EDS to only partial namespace would be there,

Reply:
> I don't yet see how consensus would work with go-header directly

Why not? go-header, afaiu, will allow the consensus node to verify a series of headers. This is similar to the mechanism behind blocksync